Custom eslint rules to help with code cleanness

danielfdsilva opened the pull request:
Split from #126.
Including the additional lint rules for further discussion.

I think these additional rules, not only improve code cleanness, but also help avoid some bugs and discouraged behaviors.

```
{
    name: 'Custom Rules',
    rules: {
      // Helps with cleaning debug statements by erroring on console.
      'no-console': 'error',
      // It's no longer needed to import React, so this just prevents weird
      // errors when you don't.
      'react/react-in-jsx-scope': 'off',
      // Array indexes as keys should not be used. The occasional time it is
      // needed, an ignore can be added.
      'react/no-array-index-key': 'error',
      // Helps with enforcing rules of hooks. Very helpful to catch wrongly
      // placed hooks, like conditional usage.
      'react-hooks/rules-of-hooks': 'error',
      // Ensure that components are PascalCase
      'react/jsx-pascal-case': 'error',
      // Force self closing components when there are no children.
      // Prevents `<MyComp prop='1'></MyComp>`
      'react/self-closing-comp': 'error',
      // Disable unused vars, handles TS-specific cases (type params,
      // interfaces) better than base rule.
      // https://typescript-eslint.io/rules/no-unused-vars/#what-benefits-does-this-rule-have-over-typescript
      '@typescript-eslint/no-unused-vars': [
        'error',
        {
          args: 'all',
          argsIgnorePattern: '^_',
          caughtErrors: 'all',
          caughtErrorsIgnorePattern: '^_',
          destructuredArrayIgnorePattern: '^_',
          varsIgnorePattern: '^_',
          ignoreRestSiblings: true
        }
      ],
      // Warn when `any` type is used. It's sometimes necessary, but should be
      // avoided when possible.
      '@typescript-eslint/no-explicit-any': 'warn'
    }
  }
```

danielfdsilva committed:
Add some eslint custom rules to help with code cleanness

AliceR commented on templates/base/eslint.config.mjs:
```diff
@@ -17,5 +17,16 @@ export default [
+  {
+    name: 'Custom Rules',
+    rules: {
+      'no-console': 2,
```
Can we use the [string](https://eslint.org/docs/latest/use/configure/rules#rule-severities) syntax to set the rules? Strings explicitly name the severity ("error", "warn", "off"), so anyone reading the config sees immediately what is intended. Also reduces cognitive load: no need to memorize the mapping.

AliceR commented on templates/base/eslint.config.mjs:
```diff
@@ -17,5 +17,16 @@ export default [
+      'no-console': 2,
+      'react/react-in-jsx-scope': 'off',
+      'react/no-array-index-key': 2,
+      'react-hooks/rules-of-hooks': 2, // Checks rules of Hooks
```
Comment seems redundant to the name of the rule. I am all for comments, but this seems a bit lost and random

AliceR reviewed (approved):
Great additions to the default rule set, thanks @danielfdsilva !
Just added two comments while going through the new rules:
- I prefer string syntax oder numbers for rules and would appreciate this to change
- Just a small cleanup comment on the comment 😜 Edit to add: You could add your explanations from the PR description as comments to the custom rules! I think it would be nice to have their existence justified for future reference!

danielfdsilva commented:
@AliceR Great suggestions!

danielfdsilva committed:
Clarify eslint rules

vgeorge commented on templates/base/eslint.config.mjs:
```diff
@@ -17,5 +17,26 @@ export default [
+      'react/jsx-pascal-case': 'error',
+      // Force self closing components when there are no children.
+      // Prevents `<MyComp prop='1'></MyComp>`
+      'react/self-closing-comp': 'error'
```
```suggestion
      // Prevents `<MyComp prop='1'></MyComp>`
      'react/self-closing-comp': 'error',
      // Disable unused vars, handles TS-specific cases (type params, interfaces) better than base rule
      '@typescript-eslint/no-unused-vars': 'error'
```

vgeorge replied:
The current configuration it displays a warning in the editor but breaks TS check. With this rule, the error would appear in the editor itself. You can test this by generating an app and place an unused variable in the code.

danielfdsilva replied:
That's a good call, but I think we should configure it to [act like ts](https://typescript-eslint.io/rules/no-unused-vars/#what-benefits-does-this-rule-have-over-typescript).

For example ts ignores rest siblings `const {a, ...rest} = props`, but eslint errors.

vgeorge replied:
@danielfdsilva what would be your proposal in this case? Adding `"no-unused-vars": "off"`? Feel free override my suggestion.

danielfdsilva replied:
@vgeorge No, not at all, just configuring the rule:
```
'@typescript-eslint/no-unused-vars': [
        'error',
        {
          args: 'all',
          argsIgnorePattern: '^_',
          caughtErrors: 'all',
          caughtErrorsIgnorePattern: '^_',
          destructuredArrayIgnorePattern: '^_',
          varsIgnorePattern: '^_',
          ignoreRestSiblings: true
        }
      ]
```

danielfdsilva replied:
In this way it ignores rest siblings, but you can also exclude variables by prefixing them with `_`. For example: `arr.map((_v, i) => i+1)`. `v` is unused but has to be declared.

vgeorge reviewed (approved):
@danielfdsilva looks good to me. Added a suggested rule.

danielfdsilva committed:
Configures stricter linting rules

Adds and configures the `@typescript-eslint/no-unused-vars` rule to improve unused variable handling, particularly in TypeScript code.  This catches more cases than the base rule.

Adds the `@typescript-eslint/no-explicit-any` rule to warn against the use of `any` types, encouraging more specific type annotations.

danielfdsilva commented:
Another rule I think is worth changing is `@typescript-eslint/no-explicit-any`. Ts doesn't complain, but eslint errors. Since sometimes is needed it works better if this is set to `warn`.

AliceR commented on templates/component-library/uswds/main.tsx:
```diff
@@ -1,5 +1,5 @@
 // eslint-disable-next-line @typescript-eslint/no-unused-vars
```
we can remove the `eslint-disable-next-line`

danielfdsilva replied:
Done!

danielfdsilva committed:
Removes unused eslint disable comment
